engine/src/main/java/org/hibernate/validator/internal/engine/tracking/PredefinedScopeProcessedBeansTrackingStrategy.java

```diff
@@ -0,0 +1,368 @@
+import org.hibernate.validator.internal.metadata.aggregated.ReturnValueMetaData;
+import org.hibernate.validator.internal.metadata.aggregated.ValidatableParametersMetaData;
+import org.hibernate.validator.internal.metadata.facets.Cascadable;
+import org.hibernate.validator.internal.properties.Signature;
+import org.hibernate.validator.internal.util.CollectionHelper;
+
+public class PredefinedScopeProcessedBeansTrackingStrategy implements ProcessedBeansTrackingStrategy {
+
+	private final Map<Class<?>, Boolean> trackingEnabledForBeans;
+
+	// TODO: signature is just name and parameters so that can clash between different beans.
+	//  with that.. do we even need to track it per signature or since
+	//  we already built the `trackingEnabledForBeans` we can just "inspect" the cascadable as we go
+	//  and check against this `trackingEnabledForBeans` to see if tracking is required ?
+	private final Map<Signature, Boolean> trackingEnabledForReturnValues;
+	private final Map<Signature, Boolean> trackingEnabledForParameters;
+
+	public PredefinedScopeProcessedBeansTrackingStrategy(Map<Class<?>, BeanMetaData<?>> rawBeanMetaDataMap) {
+		// TODO: build the maps from the information inside the beanMetaDataManager
+		//  There is a good chance we will need a structure with the whole hierarchy of constraint classes.
+		//  That's something we could add to PredefinedScopeBeanMetaDataManager, as we are already doing similar things
+		//  there (see the ClassHierarchyHelper.getHierarchy call).
+		//  In the predefined scope case, we will have the whole hierarchy of constrained classes passed to
+		//  PredefinedScopeBeanMetaDataManager.
+
+		this.trackingEnabledForBeans = CollectionHelper.toImmutableMap(
+				new TrackingEnabledStrategyBuilder( rawBeanMetaDataMap ).build()
+		);
+		this.trackingEnabledForReturnValues = CollectionHelper.toImmutableMap( new HashMap<>() );
+		this.trackingEnabledForParameters = CollectionHelper.toImmutableMap( new HashMap<>() );
+	}
+
+	private static class TrackingEnabledStrategyBuilder {
+		private final Map<Class<?>, BeanMetaData<?>> rawBeanMetaDataMap;
+		private final Map<Class<?>, Boolean> classToBeanTrackingEnabled;
+		// Map values are a set of subtypes for the key class, including "self" i.e. the "key":
+		private final Map<Class<?>, Set<Class<?>>> subtypesMap;
+
+		TrackingEnabledStrategyBuilder(Map<Class<?>, BeanMetaData<?>> rawBeanMetaDataMap) {
+			this.rawBeanMetaDataMap = rawBeanMetaDataMap;
+			this.classToBeanTrackingEnabled = CollectionHelper.newHashMap( rawBeanMetaDataMap.size() );
+			this.subtypesMap = CollectionHelper.newHashMap( rawBeanMetaDataMap.size() );
+			for ( Class<?> beanClass : rawBeanMetaDataMap.keySet() ) {
+				for ( Class<?> otherBeanClass : rawBeanMetaDataMap.keySet() ) {
+					if ( beanClass.isAssignableFrom( otherBeanClass ) ) {
+						subtypesMap.computeIfAbsent( beanClass, k -> new HashSet<>() )
+								.add( otherBeanClass );
+					}
+				}
```

Comment:
This subtype map is here to help identify subtypes that can potentially, at runtime, introduce a cycle, and if that's so, we would mark that path as requiring the bean tracking.